### DISCUSSION TITLE
Classmethods and variance

### DISCUSSION THREAD

**Discussion by decorator-factory** [microsoft/pyright#2837] [General]
It seems to me that class methods and `__init__` are exempt from variance checks :thinking: 
```py
from typing import Generic, TypeVar


T = TypeVar("T", covariant=True)


class Foo(Generic[T]):
    def __init__(self, bar: T):  # ok
        self.bar = bar

    @classmethod
    def from_first(cls, bar: tuple[T, T]) -> "Foo[T]":  # also ok
        return cls(bar[0])


class Foo2(Generic[T]):
    def __init__(self, bar: T):  # ok
        self.bar = bar

    def set_bar(self, bar: T) -> None:  # error: Covariant type variable cannot be used in parameter type
        self.bar = bar
```
I do realize now that it would be impossible to even make an `__init__` for a generic class if this was not the case. Are there any additional details of how this works? I wasn't able to find anything about it in PEP 484

**Comment by decorator-factory** [comment]
Hmm, this is strange. 
```py
class Foo(Generic[T]):
    @classmethod
    def make1(cls, bar: T) -> "Foo[T]":  # error: Covariant type variable cannot be used in parameter type
        return cls(bar)
        
    @classmethod
    def make2(cls, bar: tuple[T]) -> "Foo[T]":  # ok
        return cls(bar[0])
```
Is this as designed?

**Reply by erictraut** [reply]
Yes, this is as designed. Mypy gives the same error message.

**Reply by decorator-factory** [reply]
So `T` is not allowed but `tuple[T]` is allowed? That doesn't make sense to me

**Reply by erictraut** [reply]
The variance of `T` in the expression `tuple[T]` is dictated by the class `tuple`. The variance of a "raw" `T` is dictated by the context in which it is used.

**Reply by decorator-factory** [reply]
Sorry, I don't understand. Could you perhaps give an example of why it has to be like this?

From what I understand, if `T` is covariant, and `A` is a subtype of `B`, then `tuple[A]` is a subtype of `tuple[B]`.
So if the rule prohibits a call like this:
```py
a: A
foo_b = Foo[B].make1(a)
```
why does it allow this, which is the same?
```py
a: A
ta = (a,)
tb: tuple[B] = ta
foo_b = Foo[B].make2(tb)
```
